ydb/core/scheme/scheme_tablecell_ut.cpp

```diff
@@ -193,13 +193,68 @@ Y_UNIT_TEST_SUITE(Scheme) {
         }
     }
 
+    ui32 GetCellHeaderRawValueOld(ui32 cellSize, bool isNull) {
+        struct TValue {
+            ui32 Size : 31;
+            ui32 IsNull : 1;
+        };
+
+        TValue value;
+        value.Size = cellSize;
+        value.IsNull = isNull;
+
+        ui32 rawValue = 0;
+        memcpy(&rawValue, &value, sizeof(ui32));
+
+        return rawValue;
+    }
+
+    ui32 GetCellHeaderRawValueNew(ui32 cellSize, bool isNull) {
+        struct TCellHeader {
+            TCellHeader() = default;
+
+            TCellHeader(ui32 rawValue)
+                : RawValue(rawValue)
+            {}
+
+            TCellHeader(ui32 cellSize, bool isNull)
+                : RawValue(cellSize | (static_cast<ui32>(isNull) << 31ULL))
+            {}
+
+            ui32 CellSize() const {
+                return RawValue & ~(1ULL << 31);
+            }
+
+            bool IsNull() const {
+                return RawValue & (1ULL << 31);
+            };
+
+            ui32 RawValue = 0;
+        };
+
+        TCellHeader cellHeader(cellSize, isNull);
+        UNIT_ASSERT_EQUAL(cellHeader.CellSize(), cellSize);
+        UNIT_ASSERT_EQUAL(cellHeader.IsNull(), isNull);
+
+        return cellHeader.RawValue;
+    }
+
+    Y_UNIT_TEST(TSerializedCellVecCellHeader) {
+        std::vector<size_t> sizes = {0, 1, 10, 20, 50, 100, 200, 500, 1000, 1500, 2000, 5000, std::numeric_limits<ui32>::max() >> 1};
+        for (size_t size : sizes)
+        {
+            UNIT_ASSERT_EQUAL(GetCellHeaderRawValueOld(size, true), GetCellHeaderRawValueNew(size, true));
```

Comment:
You're not testing any real code here. If you want to make sure `TCellHeader` in scheme_tablecell.cpp behaves a certain way, you should probably make that class publicly available and tests its results. The result of `TCellHeader` is just math, so result can be compared with a precomputed expected value (so if serialization changes test would fail instead passing because two changes cancelled each other). The result of bitfields could be verified using memcpy or WriteUnaligned/ReadUnaligned using some buffer and again compared with the expected ui32 value, although it's not clear why, new code is not using it anyway. Even better if tests would have some preserialized examples that we expect to deserialize a certain way (so it would fail on big endian, since current format is not portable between little and big endian).

Testing round trip in binary serialization is often not enough to catch compatibility issues and changes.

Reply:
Updated.